Deprecate `to_s(format)` in favor of `to_formatted_s(format)`

rafaelfranca opened the pull request:
Ruby 3.1 is going to introduce an [optimization][] that makes interpolation of some types of objects faster, unless there is a custom implementation of to_s. Since Rails is overriding `to_s` for a bunch of core classes it means that this optimization in Rails applications will be disabled.

In order to allow Rails applications to use this optimization in the future we are deprecating all the reasons we override `to_s` in those core classes in favor of using `to_formatted_s`.

I'm planning another commit to configure if applications want the `to_s` override or not in order to allow people to already get Rails 7.0 taking advantage of the Ruby 3.1 optimization.

Long term plan is to remove the `to_s` override of all those core types, even though the Ruby optimization only works for some of them.

[optimization]: https://github.com/ruby/ruby/commit/b08dacfea39ad8da3f1fd7fdd0e4538cc892ec44

rafaelfranca committed:
Use `to_formatted_s(:db)` instead of `to_s(:db)` internally

Ruby 3.1 introduced an optimization to string interpolation for some
core classes in https://github.com/ruby/ruby/commit/b08dacfea39ad8da3f1fd7fdd0e4538cc892ec44.

But since we override `to_s` in some of those core classes to add behavior
like `to_s(:db)`, all Rails applications will not be able to take advantage
of that improvement.

Since we can use the `to_formatted_s` alias for the Rails specific behavior
it is best for us to deprecate the `to_s` core extension and allow Rails
applications to get the proformace improvement.

This commit starts removing all the `to_s(:db)` calls inside the framework
so we can deprecate the core extension in the next commit.

rafaelfranca committed:
Deprecate `to_s(format)` in favor of `to_formatted_s(format)`

Ruby 3.1 is going to introduce an [optimization][] that makes interpolation
of some types of objects faster, unless there is a custom implementation
of to_s. Since Rails is overriding `to_s` for a bunch of core classes it
means that this optimization in Rails applications will be disabled.

In order to allow Rails applications to use this optimization in
the future we are deprecating all the reasons we override `to_s` in
those core classes in favor of using `to_formatted_s`.

[optimization]: https://github.com/ruby/ruby/commit/b08dacfea39ad8da3f1fd7fdd0e4538cc892ec44

jonathanhefner reviewed (commented):
Just a few find / replace glitches. :+1:

rafaelfranca committed:
Alias to_formatted_s as to_fs

rafaelfranca committed:
Add config to disable the `to_s` override in Ruby core classes

This will allow users to take advantage early of Ruby 3.1 optimization
in string interpolation.

rafaelfranca committed:
Fix typo
